### PR TITLE
Remove `stat` and `errmsg` args from `caf` collectives interfaces

### LINKED ISSUE #81: Remove unnecessary argument definitions in collectives
Until this issue is handled, any collective subroutine may crash.   For now, Caffeine will not provide error messages or statuses in the event of the failure of a collective subroutine:
- [x] Remove any handling of `errmsg` and `errmsg_alloc` from all collective subroutines
- [x] Set `stat=0` in all collective subroutines
- [x] Remove all references to unused variables in the C interfaces
- [x] Ensure all optional arguments in the prif collectives are checked if present before referencing them, unless passing them on to other optional arguments in a routine call

### PR DESCRIPTION
**Pull request by ktras**
Closes #81.

### CODE REVIEW

File: src/caffeine/caffeine.c [hunk 1]
```diff
@@ -183,9 +179,11 @@ static void set_stat_errmsg_or_abort(int* stat, char* errmsg, const int return_s
   }
 }
 
-void caf_co_max(CFI_cdesc_t* a_desc, int result_image, int* stat, char* errmsg, size_t num_elements, gex_TM_t team)
+void caf_co_max(CFI_cdesc_t* a_desc, int result_image, size_t num_elements, gex_TM_t team)
 {
   gex_DT_t a_type;
+  int* stat = NULL;
+  char* errmsg = NULL;
```
[the remark this answers is not in the document]
Comment (ktras) [reply src/caffeine/caffeine.c:186]: @everythingfunctional They are used in calls to the function `set_stat_errmsg_or_abort` if the type of the coarray is unsupported. Whether this is still necessary is another question that we should discuss. I meant to add this in the discussion of the PR but forget. (This is the same for the other local declarations added that you commented on).

File: src/caffeine/caffeine.c [hunk 2]
```diff
@@ -239,13 +237,13 @@ void caf_co_min(CFI_cdesc_t* a_desc, int result_image, int* stat, char* errmsg,
...
-
-  if (stat != NULL) *stat = 0;
 }
 
-void caf_co_sum(CFI_cdesc_t* a_desc, int result_image, int* stat, char* errmsg, size_t num_elements, gex_TM_t team)
+void caf_co_sum(CFI_cdesc_t* a_desc, int result_image, size_t num_elements, gex_TM_t team)
 {
   gex_DT_t a_type;
+  int* stat = NULL;
+  char* errmsg = NULL;
```
[the remark this answers is not in the document]
Comment (ktras) [reply src/caffeine/caffeine.c:246]: See above reply to comment.

File: src/caffeine/caffeine.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (everythingfunctional) [reply src/caffeine/caffeine.c:186]: Was looking too fast and didn't see that. You could just pass `NULL` for now.

[the remark this answers is not in the document]
Comment (ktras) [reply src/caffeine/caffeine.c]: True. I have just made that change. Thanks.

File: src/caffeine/caffeine.c [hunk 3]
```diff
@@ -194,7 +190,7 @@ void caf_co_max(CFI_cdesc_t* a_desc, int result_image, int* stat, char* errmsg,
     case CFI_type_float:            a_type = GEX_DT_FLT; break;
     case CFI_type_double:           a_type = GEX_DT_DBL; break;
     default:
-      set_stat_errmsg_or_abort(stat, errmsg, UNRECOGNIZED_TYPE, "");
+      set_stat_errmsg_or_abort(NULL, NULL, UNRECOGNIZED_TYPE, "");
```
Comment (bonachea) [inline src/caffeine/caffeine.c:193]: IIUC the call above is equivalent to `gasnett_fatalerror("")`, which is better than `abort()` but still not great: the message doesn't give a hint about what's wrong, due to the level of call indirection the failing line number won't even point to this line as the root cause, nor does it give a hint about what unrecognized type was passed.

```suggestion
      gasnett_fatalerror("Unrecognized type: %d", (int)a_desc->type);
```

File: src/caffeine/caffeine.c [hunk 4]
```diff
@@ -224,7 +218,7 @@ void caf_co_min(CFI_cdesc_t* a_desc, int result_image, int* stat, char* errmsg,
...
```
Comment (bonachea) [inline src/caffeine/caffeine.c:221]:
```suggestion
      gasnett_fatalerror("Unrecognized type: %d", (int)a_desc->type);
```

File: src/caffeine/caffeine.c [hunk 5]
```diff
@@ -258,7 +250,7 @@ void caf_co_sum(CFI_cdesc_t* a_desc, int result_image, int* stat, char* errmsg,
     case float_Complex_workaround:  a_type = GEX_DT_FLT; num_elements *= 2; c_sizeof_a /= 2; break;
     case double_Complex_workaround: a_type = GEX_DT_DBL; num_elements *= 2; c_sizeof_a /= 2; break;
     default:
-      set_stat_errmsg_or_abort(stat, errmsg, UNRECOGNIZED_TYPE, "");
+      set_stat_errmsg_or_abort(NULL, NULL, UNRECOGNIZED_TYPE, "");
```
Comment (bonachea) [inline src/caffeine/caffeine.c:253]:
```suggestion
      gasnett_fatalerror("Unrecognized type: %d", (int)a_desc->type);
```

File: src/caffeine/collective_subroutines/co_max_s.f90 [hunk 6]
```diff
@@ -1,32 +1,21 @@
...
-
       call caf_co_max( &
-          a, optional_value(result_image), stat_c_ptr, errmsg_c_ptr, int(product(shape(a)), c_size_t), current_team%gex_team)
-
-      call c_f_pointer(errmsg_c_ptr, errmsg_f_ptr) ! no need to do this for stat was passed by reference
-      errmsg = errmsg_f_ptr ! copy the output back & truncate the null terminator
-
+          a, optional_value(result_image), int(product(shape(a)), c_size_t), current_team%gex_team)
     else if (caf_is_f_string(a)) then
       call prif_co_reduce(a, c_funloc(reverse_alphabetize), optional_value(result_image), stat, errmsg)
```
Comment (bonachea) [inline src/caffeine/collective_subroutines/co_max_s.f90:20]: `errmsg_alloc` is missing here.

```suggestion
      call prif_co_reduce(a, c_funloc(reverse_alphabetize), optional_value(result_image), stat, errmsg, errmsg_alloc)
```

File: src/caffeine/collective_subroutines/co_min_s.f90 [hunk 7]
```diff
@@ -1,32 +1,21 @@
...
-
       call caf_co_min( &
-          a, optional_value(result_image), stat_c_ptr, errmsg_c_ptr, int(product(shape(a)), c_size_t), current_team%gex_team)
-
-      call c_f_pointer(errmsg_c_ptr, errmsg_f_ptr) ! no need to do this for stat was passed by reference
-      errmsg = errmsg_f_ptr ! copy the output back & truncate the null terminator
-
+          a, optional_value(result_image), int(product(shape(a)), c_size_t), current_team%gex_team)
     else if (caf_is_f_string(a)) then
       call prif_co_reduce(a, c_funloc(alphabetize), optional_value(result_image), stat, errmsg)
```
Comment (bonachea) [inline src/caffeine/collective_subroutines/co_min_s.f90:20]: `errmsg_alloc` is missing here.
```suggestion
      call prif_co_reduce(a, c_funloc(alphabetize), optional_value(result_image), stat, errmsg, errmsg_alloc)
```

**Review by bonachea — CHANGES_REQUESTED** [review]
Made a few minor requests, otherwise LGTM

**Comment by ktras** [comment]
@bonachea Thanks for the suggested changes. After making them, the function `set_stat_errmsg_or_abort` is never called. Is this a function we think will likely be useful in the future? Or should we delete the function? @rouson, @everythingfunctional, any thoughts?

**Comment by everythingfunctional** [comment]
I would have changed `set_stat_errmsg_or_abort` to call `gasnett_fatalerror` instead of hard-coding that multiple places.

**Comment by bonachea** [comment]
> I would have changed `set_stat_errmsg_or_abort` to call `gasnett_fatalerror` instead of hard-coding that multiple places.

@everythingfunctional what you suggest is essentially the "before" code. The "after" code I suggested with a direct call will generate a more useful error message.

**Comment by bonachea** [comment]
> @bonachea Thanks for the suggested changes. After making them, the function `set_stat_errmsg_or_abort` is never called. Is this a function we think will likely be useful in the future? Or should we delete the function? @rouson, @everythingfunctional, any thoughts?

@ktras I'm in favor of removing the dead code. I don't expect we'll be supporting any non-fatal errors anytime soon other than allocation failure, and that doesn't use this code path.

**Comment by everythingfunctional** [comment]
Ok, but then we should adjust the interface to `set_stat_errmsg_or_abort` to allow for passing the information/message we want, rather than getting rid of it. At some point we will want the logic it describes, and will be easier to make the change in one place instead of multiple.

That said, if you think

> I don't expect we'll be supporting any non-fatal errors anytime soon other than allocation failure, and that doesn't use this code path.

I'm fine with just delaying that refactoring for now.

**Review by bonachea — APPROVED** [review]

**Pull request merged**